.Net: Updates for ChatMessageContent to support complex content model

dmytrostruk opened the pull request:
### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

Resolves: https://github.com/microsoft/semantic-kernel/issues/3874
Resolves: https://github.com/microsoft/semantic-kernel/issues/3522
Resolves: #3768 

This PR contains changes to prepare public API of `ChatMessageContent` to support extensible content model in the future. With latest OpenAI API it's possible to have content as `string` or `array` of content objects (scenario for GPT vision):
![image](https://github.com/microsoft/semantic-kernel/assets/13853051/53817e93-1cdc-4b81-aab9-0d6369e9f1e6)

### Description

<!-- Describe your changes, the overall approach, the underlying design.
     These notes will help understanding how your code works. Thanks! -->

1. Added `ChatMessageContentItemCollection` to store multiple contents of different types in chat message.
2. Added `ChatMessageContentItemCollection? Items` property to `ChatMessageContent`, made `string? Content` property nullable.
3. Added `ImageContent` class that stores image URI.
4. Added ADR and diagram of chat and text content models.
5. Added `Example74_GPTVision` to demonstrate the functionality with GPT Vision model.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

markwallace-microsoft commented on dotnet/src/SemanticKernel.Abstractions/AI/TextToImage/ImageContent.cs:
```diff
@@ -0,0 +1,43 @@
+/// <summary>
+/// Represents image content.
+/// </summary>
+public class ImageContent : ContentBase
```
This should be sealed. If a developer was a customized `ImageContent` they should extend `ContentBase` directly.

markwallace-microsoft reviewed (approved):
One small issue

SergeyMenshykh commented on dotnet/src/SemanticKernel.Abstractions/AI/TextToImage/ImageContent.cs:
```diff
@@ -0,0 +1,43 @@
+    /// <summary>
+    /// The URI of image.
+    /// </summary>
+    public Uri Uri { get; set; }
```
Sooner or later, we will need to support images represented as byte arrays. It would be good to consider whether this will be modeled either through a new `byte[] Bytes {get; set;}` property of the class or as an `ImageByteContent` subclass of `ContentBase`. This exercise will be useful in shaping the SK public API surface now to prevent potential breaking changes in the future.

RogerBarreto replied:
## Points to consider 

How are we going to distinct referenced contents from non referenced ones?

### Following the abstraction path: 

`StreamingContentBase` has a `abstract byte[] ToByteArray()` method.

We might consider bringing this to a `ByteContent` abstract class or directly to the `ContentBase` to be similar.

### How to handle Referenced/Link/Anchor contents?

Any content could be referenced by a Uri, might worth having a `ILinkContent` that ImageContent/AudioContent, etc could inherit or implement

```csharp
interface ILinkContent {
    Uri Uri { get }
}

public OpenAIReferencedImageContent : ContentBase, IReferencedContent {
    Uri Uri { get; set; }
}
```

dmytrostruk replied:
> Sooner or later, we will need to support images represented as byte arrays. It would be good to consider whether this will be modeled either through a new `byte[] Bytes {get; set;}` property of the class or as an `ImageByteContent` subclass of `ContentBase`. This exercise will be useful in shaping the SK public API surface now to prevent potential breaking changes in the future.

Both options (new property or new separate class) look like addition, so I'm not sure there will be breaking changes when we want to extend it. In case of new property, we will have to convert `Uri` property to nullable, because it will be either image link or image byte array. 

> Any content could be referenced by a Uri, might worth having a `ILinkContent` that ImageContent/AudioContent, etc could inherit or implement

Hmm, that's interesting idea. At the moment, I can't find a scenario when `ILinkContent` abstraction will be useful for us or end-users, could you share any examples?

SergeyMenshykh replied:
> In case of new property, we will have to convert Uri property to nullable.

Making a property nullable can be considered a breaking change, right? Particularly if the existing consumers are not expecting or handling a null value for the property. Should we make it nullable now to be prepared?

dmytrostruk replied:
@SergeyMenshykh Yes, that makes sense. Converted `Uri` parameter to nullable.

SergeyMenshykh commented on dotnet/samples/KernelSyntaxExamples/Example49_LogitBias.cs:
```diff
@@ -44,14 +44,14 @@ public static async Task RunAsync()
 
         var replyMessage = await chatCompletionService.GetChatMessageContentAsync(chatHistory, settings);
-        chatHistory.AddAssistantMessage(replyMessage.Content);
+        chatHistory.AddAssistantMessage(replyMessage.Content!);
```
Why are we certain (using the `!` operator) that the content is not null here?

dmytrostruk replied:
That's a good question. Currently, we operate with `ChatMessageContent` as model for request and response at the same time. In case of request, property `Content` can be used for plain text or property `Items` for multi-modal request. While for response, only `Content` property is used. That's why `Content` property is nullable, because for multi-modal request it's not required.. I think we can either use different classes for chat message request and response scenario or remove `Content` property completely and use only `Items` in request and response. In this case, we will avoid null checks. Although, with `Items`, users will have to do `Items[0]` to get response every time, so we will have to provide additional methods for better usability.

In this specific example, we know that `Content` is not null, because we set a response from AI into this property, and in case of some issue we just throw an exception.

RogerBarreto commented on dotnet/src/SemanticKernel.Abstractions/AI/TextToImage/ImageContent.cs:
```diff
@@ -0,0 +1,43 @@
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.SemanticKernel.TextToImage;
```
Out of scope of this PR, but I think would be useful to move all the content abstractions to the root namespace.

- ContentBase
- ChatMessageContent
- TextContent
- ImageContent

Streaming also.

```suggestion
namespace Microsoft.SemanticKernel;
```

dmytrostruk replied:
Thanks, I think it's a good idea. In case we make a decision to do that, I will create a separate issue to avoid big number of changes in this PR.

dmytrostruk commented on dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:
```diff
@@ -664,6 +665,27 @@ private static ChatRequestMessage GetRequestMessage(ChatRole chatRole, string co
+    {
+        if (message.Items is { Count: > 0 })
+        {
+            var contentItems = message.Items.Select(GetChatMessageContentItem).ToList();
```
Yes, thank you. Forgot to remove it after initial version of this implementation.

stephentoub commented:
What's the relationship now between ChatMessageContentItemCollection and ChatHistory? They're both a collection of chat message contents, right?

dmytrostruk commented:
> They're both a collection of chat message contents, right?

No, `ChatMessageContentItemCollection` is a collection of `ContentBase` items, like `TextContent`, `ImageContent` or `ChatMessageContent` itself, while `ChatHistory` is a collection of concrete type `ChatMessageContent`. Yes, it looks like `ChatMessageContentItemCollection` can be re-used in the same way as `ChatHistory`, although both classes have different responsibilities. 

In our case, `ChatHistory` is a main container for all chat messages with methods like `AddUserMessage`, `AddSystemMessage` etc, while `ChatMessageContentItemCollection` is a collection of chat message contents, which can be text or image url for example.

stephentoub commented:
> No, ChatMessageContentItemCollection is a collection of ContentBase items, like TextContent, ImageContent or ChatMessageContent itself, while ChatHistory is a collection of concrete type ChatMessageContent. 

Right, that's my point: ChatHistory _is_ a collection of ContentBase items, just guaranteeing a more specialized ContentBase-derived type. So should that relationship be modeled in the API, so that a ChatHistory can be used anywhere a ChatMessageContentItemCollection is employed?

dmytrostruk commented:
> So should that relationship be modeled in the API, so that a ChatHistory can be used anywhere a ChatMessageContentItemCollection is employed?

I think the reason why it's confusing is because we use `TextContent` or `ChatMessageContent` models as response from AI and at the same time we use `TextContent` inside `ChatMessageContent` as part of chat message text content. Maybe it's not a problem, but probably that's the reason why hierarchy is not clear. (That's why I've added class diagram to this PR :) ).

From user perspective, I would like to clearly differentiate between `ChatHistory` from `ChatMessageContentItemCollection`. In initial `ChatMessageContentItemCollection` I've added a check that `ChatMessageContent` (as `ContentBase`) can't be added to the collection (unline in `ChatHistory`), because it doesn't make sense to include chat message as content of chat message. But I removed that check, because this restriction doesn't actually provide any benefits, and on top of that, checking `ContentBase` type every time on adding new item to the collection may be costly.

Answering your question, I'm not sure if using ChatHistory anywhere a ChatMessageContentItemCollection is employed will provide any benefits, but maybe I'm missing something. Some use-cases or scenarios will definitely help.
